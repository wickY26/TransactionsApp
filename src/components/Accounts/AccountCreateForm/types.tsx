import { Account } from '../../../containers/Accounts/types';

export interface AccountCreateFormProps {
  onCreateAccount: (data: Account) => void;
}
