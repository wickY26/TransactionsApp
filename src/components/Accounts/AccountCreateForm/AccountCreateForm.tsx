import React, { FC, FormEvent, useState } from 'react';
import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import SaveIcon from '@material-ui/icons/Save';

import { AccountCreateFormProps } from './types';

const AccountCreateForm: FC<AccountCreateFormProps> = ({ onCreateAccount }) => {
  const [name, setName] = useState<string>('');
  const [surname, setSurname] = useState<string>('');

  const handleSubmit = (evt: FormEvent<HTMLFormElement>) => {
      evt.preventDefault();
      onCreateAccount({ name, surname });
      setName('');
      setSurname('');
  }

  return (
    <form noValidate autoComplete="off" onSubmit={handleSubmit}>
      <Box display="flex" flexDirection="column" m={2}>
        <TextField
          inputProps={{ "data-testid": "name" }}
          margin="normal"
          label="Name" 
          value={name}
          onChange={e => setName(e.target.value)}/>
        <TextField
          inputProps={{ "data-testid": "surname" }}
          margin="normal"
          label="Surname" 
          value={surname}
          onChange={e => setSurname(e.target.value)}/>
        <Button
          data-testid="submitBtn"
          variant="contained"
          color="primary"
          startIcon={<SaveIcon />}
          type="submit"
          disabled={!name || !surname}
        >
          Create
        </Button>
      </Box>
    </form>
  );
}

export default AccountCreateForm;
