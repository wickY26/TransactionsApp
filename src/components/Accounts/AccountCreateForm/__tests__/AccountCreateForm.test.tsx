import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import AccountCreateForm from '..';

const mockOnCreateAccount = jest.fn();

const renderComponent = () => render(
  <AccountCreateForm onCreateAccount={mockOnCreateAccount} />
);

describe('AccountCreateForm', () => {

  it('should render correctly', () => {
    const component = renderComponent();
    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should make submit button disabled when name or surname fields empty', () => {
    renderComponent();

    expect(screen.getByTestId('submitBtn')).toHaveAttribute('disabled');
  });

  it('should call onCreateAccount when form is submitted', () => {
    renderComponent();

    fireEvent.change(screen.getByTestId('name'), { target: { value: 'Poyraz' } });
    fireEvent.change(screen.getByTestId('surname'), { target: { value: 'Yilmaz' } });
    fireEvent.click(screen.getByTestId('submitBtn'));

    expect(mockOnCreateAccount).toHaveBeenCalled();
    expect(mockOnCreateAccount).toHaveBeenCalledWith({name:'Poyraz', surname: 'Yilmaz'});
  });

  it('should reset form after submit', () => {
    renderComponent();
    const nameField = screen.getByTestId('name') as HTMLInputElement;
    const surnameField = screen.getByTestId('surname') as HTMLInputElement;

    fireEvent.change(nameField, { target: { value: 'Poyraz' } });
    fireEvent.change(surnameField, { target: { value: 'Yilmaz' } });
    fireEvent.click(screen.getByTestId('submitBtn'));

    expect(nameField.value).toEqual('');
    expect(surnameField.value).toEqual('');
  });

});
