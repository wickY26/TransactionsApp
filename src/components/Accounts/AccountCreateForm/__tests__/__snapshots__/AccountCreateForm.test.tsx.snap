// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountCreateForm should render correctly 1`] = `
<DocumentFragment>
  <form
    autocomplete="off"
    novalidate=""
  >
    <div
      class="MuiBox-root MuiBox-root-1"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
        >
          Name
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            data-testid="name"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
        >
          Surname
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            data-testid="surname"
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
        data-testid="submitBtn"
        disabled=""
        tabindex="-1"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
              />
            </svg>
          </span>
          Create
        </span>
      </button>
    </div>
  </form>
</DocumentFragment>
`;
